### DISCUSSION TITLE
Custom flag/commands via pnpmfile

### DISCUSSION THREAD

**Discussion by vjpr** [pnpm/pnpm#3179] [Ideas]
I was thinking that a lot of the filtering cli syntax can become quite complicated, and is sometimes restrictive if it doesn't consider all use cases. The best solution might not become apparent until later.

So what about we provide a rich api in the `pnpmfile.js` for filtering (and maybe other things), and then we can allow custom flags and commands to be added to the cli.

```
pnpm custom-command --x-custom-flag

pnpm install --x-custom-flag
```

---

Something like this:-

```
module.exports = {
  filterWorkspace,
}

// Return the packages to consider during `pnpm mutli link` process.
function filterWorkspace({workspaces, customFlags}) {
  customFlags['custom-flag']
}
```

---

## Why?

- Allow experimenting with different commands and flags before rolling them out and making the CLI api fixed.

## Why not a wrapper?

You can interact with different stages of the install.

**Comment by zkochan** [comment]
Seems complicated. What if the recursive commands would support streaming in the list of packages?

```
cat list-of-packages | pnpm recursive install
```
